test_sanity_arith_expr.cpp

```diff
@@ -146,6 +146,147 @@ void TestInterestingIntegerParams(std::function<std::function<void(T, T)>()> fnG
     }
 }
 
+template<typename T, typename U>
+void AdditionWithDifferentTypesHelper(T lhs, U rhs)
+{
+    AutoThreadPochiVMContext apv;
+    AutoThreadErrorContext arc;
+    AutoThreadLLVMCodegenContext alc;
+    NewModule("test");
+    using FnPrototype = decltype(lhs + rhs) (*)(T, U);
+    auto [fn, a, b] = NewFunction<FnPrototype>("MyFn");
+    fn.SetBody(
+        Return(a + b)
+    );
+    ReleaseAssert(thread_pochiVMContext->m_curModule->Validate());
+    thread_pochiVMContext->m_curModule->PrepareForDebugInterp();
+    thread_pochiVMContext->m_curModule->PrepareForFastInterp();
+    auto interpFn = thread_pochiVMContext->m_curModule->
+                           GetDebugInterpGeneratedFunction<FnPrototype>("MyFn");
+    ReleaseAssert(interpFn);
+    auto fastinterpFn = thread_pochiVMContext->m_curModule->
+                           GetFastInterpGeneratedFunction<FnPrototype>("MyFn");
+    ReleaseAssert(fastinterpFn);                                                 
+    thread_pochiVMContext->m_curModule->EmitIR();
+    thread_pochiVMContext->m_curModule->OptimizeIRIfNotDebugMode(2 /*optLevel*/);
+    SimpleJIT jit;
+    jit.SetModule(thread_pochiVMContext->m_curModule);
+    FnPrototype jitFn = jit.GetFunction<FnPrototype>("MyFn");
+    std::vector<decltype(jitFn(lhs, rhs))> answers = {jitFn(lhs, rhs), interpFn(lhs, rhs), fastinterpFn(lhs, rhs)};
+
+    // Ignore warnings about implicit float/double conversions/promotions
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wimplicit-int-float-conversion"
+    #pragma clang diagnostic ignored "-Wdouble-promotion"
+
+    auto expected = lhs + rhs;
```

Comment:
I'm a bit suspicious if this would work correctly. If I recall it correctly C++ has the weird thing that promotes everything <32bits to 32 bits (which Pochi currently does not support right now: if you add two uint8_t together you still get a uint8_t, not a uint32_t). So this will likely fail in overflow cases if you add, say, a uint8_t with a uint16_t (e.g 1 + 65535), but I'm not sure.

Reply:
This test doesn't check 8 bit and 16 bit integer types since only the 32/64 bit ones are implemented right now. So I don't think this behavior should affect this function. I added an assert to ensure that this test function isn't run on int8_t/int16_t.